gitolize.sh

```diff
@@ -0,0 +1,171 @@
+#!/bin/bash
+
+# A script wrapping a shell command so that it runs with a directory populated from a git repository.
+# In the write mode, the repository is locked using a temporary branch on a per-project basis and
+# changes made in that directory by the shell command get pushed back to the repository.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+usage() {
+  echo "Usage: $0 [-b <branch>] [-l <local_directory>] [-m message] [-p project] [-r repository] [-v] [-w] [command...]" 1>&2
+  exit 1
+}
+
+git() {
+  COMMAND="$1"
+  shift
+  EXTRA_FLAGS=
+  case "$COMMAND" in
+    add)
+      EXTRA_FLAGS="$GIT_VERBOSE_FLAG"
+      ;;
+    *)
+      EXTRA_FLAGS="$GIT_QUIET_FLAG"
+      ;;
+  esac
+  if [[ "$VERBOSE" ]]
+  then
+    echo "running $GIT -C "$LOCAL_DIRECTORY" $COMMAND $EXTRA_FLAGS $@" 1>&2
+  fi
+  $GIT -C "$LOCAL_DIRECTORY" $COMMAND $EXTRA_FLAGS "$@"
+}
+
+LOCK_FILE=gitolize.lock
+
+EXIT_CODE=0
+
+GIT="$(which git)"
+GIT_QUIET_FLAG=--quiet
+GIT_VERBOSE_FLAG=
+
+BRANCH=main
+LOCAL_DIRECTORY=
+CLEANUP_LOCAL_DIRECTORY=
+COMMIT_MESSAGE=
+PROJECT=
+GIT_REPOSITORY=
+VERBOSE=
+WRITE=
+
+
+while getopts b:l:m:p:r:vw OPTION
+do
+  case $OPTION in
+    b)
+      BRANCH="$OPTARG"
+      ;;
+    l)
+      LOCAL_DIRECTORY="$OPTARG"
+      ;;
+    m)
+      COMMIT_MESSAGE="$OPTARG"
+      ;;
+    p)
+      PROJECT="$OPTARG"
+      ;;
+    r)
+      GIT_REPOSITORY="$OPTARG"
+      ;;
+    v)
+      VERBOSE=true
+      ;;
+    w)
+      WRITE=true
+      ;;
+    *)
+      usage
+      ;;
+  esac
+done
+
+shift $((OPTIND-1))
+
+if [[ "$#" < 1 ]]
+then
+  usage
+fi
+
+if [[ ! "$LOCAL_DIRECTORY" ]]
+then
+  LOCAL_DIRECTORY="$(mktemp -d)"
+  CLEANUP_LOCAL_DIRECTORY=true
+fi
+
+if [[ "$PROJECT" ]]
+then
+   LOCK_BRANCH="lock/${PROJECT}"
+else
+   LOCK_BRANCH="lock"
+fi
+
+if [[ "$VERBOSE" ]]
+then
+  GIT_QUIET_FLAG=
+  GIT_VERBOSE_FLAG=--verbose
+fi
+
+if [[ ! "$COMMIT_MESSAGE" ]]
+then
+  COMMIT_MESSAGE="$*"
+fi
+
+if [[ -d "$LOCAL_DIRECTORY/.git" ]]
+then
+  git checkout "$BRANCH"
+  git pull
+else
+  git clone \
+    --branch "$BRANCH" \
+    --depth 1 \
+    "$GIT_REPOSITORY" \
+    "$LOCAL_DIRECTORY"
+fi
+
+if [[ "$WRITE" ]]
+then
+  git checkout -b "$LOCK_BRANCH"
+  LOCK_MESSAGE="Locking for $COMMIT_MESSAGE"
+  echo "$LOCK_MESSAGE" >> "$LOCAL_DIRECTORY/$LOCK_FILE"
+  git add "$LOCK_FILE"
+  git commit --message "$LOCK_MESSAGE"
+  git push --set-upstream origin "$LOCK_BRANCH" || {
+    EXIT_CODE=$?
+    echo "locking failed" 1>&2
+    exit "$EXIT_CODE"
+  }
+  git checkout "$BRANCH"
+fi
+
+GITOLIZE_DIRECTORY="$LOCAL_DIRECTORY" "$@"
+EXIT_CODE=$?
+if [[ "$EXIT_CODE" != 0 ]]
+then
+  CLEANUP_LOCAL_DIRECTORY=
+  if [[ "$VERBOSE" ]]
+  then
+    echo "failed with exit code $EXIT_CODE: $@" 1>&2
+  fi
+fi
+
+if [[ "$WRITE" ]]
+then
+  git add .
+  git commit --allow-empty --message "$COMMIT_MESSAGE"
+  if ! git push 2>/dev/null
+  then
+    # in case something has been written for another project in the meantime
+    git pull --rebase
```

Comment:
This is a nice best-effort attempt to still push the changes but I'm wondering how we could still improve this situation in cases where this attempt won't be sufficient (e.g. in case of repositories with frequent pushes). If this is run in a CI environment, the changes made by the gitolized command will be lost. I would suggest the following improvements:
- document this behaviour so the users can protect from such situations (e.g. taking additional steps in their CI workflows and backup the changes using other means - as build artifacts, using backup storage etc)
- create yet another branch to push these changes so they are not lost (e.g. with a branch name set to a random UUID)

WDYT?

Reply:
Went with `$BRANCH-$SHA` for the yet another branch name in 0a13c5fead4e94808393dcdb6cae3f4126164fd5, WDYT?

Reply:
LGTM.